File: 2021/09/13/MatrixReshape.java
/**
 * https://leetcode.com/problems/reshape-the-matrix/
 * @author Andrew Jarombek
 * @since 9/13/2021
 */

public class MatrixReshape {
    public int[][] matrixReshape(int[][] mat, int r, int c) {
        if (r * c != mat.length * mat[0].length) {
            return mat;
        }

        int[][] newMat = new int[r][c];

        int row = 0;
        int col = 0;

        for (int i = 0; i < mat.length; i++) {
            for (int j = 0; j < mat[i].length; j++) {
                newMat[row][col] = mat[i][j];

                if (col + 1 == c) {
                    row++;
                    col = 0;
                } else {
                    col++;
                }
            }
        }

        return newMat;
    }
}

File: 2022/09/13/matrixReshape.js
/**
 * https://leetcode.com/problems/reshape-the-matrix/
 * @author Andrew Jarombek
 * @since 9/13/2022
 */

const matrixReshape = function(mat, r, c) {
    const rows = mat.length;
    const cols = mat[0].length;

    if (r * c !== rows * cols)
        return mat;

    const result = [];
    let index = 0;

    for (let i = 0; i < r; i++) {
        const row = [];

        for (let j = 0; j < c; j++) {
            row.push(mat[Math.floor(index / cols)][index % cols]);
            index++;
        }

        result.push(row);
    }

    return result;
}

let result = matrixReshape([[1, 2], [3, 4]], 1, 4);
let expected = [[1, 2, 3, 4]]
console.assert(expected.every(
    (row, r) =>
        row.every((item, c) =>
            item === result[r][c])
));

result = matrixReshape([[1, 2], [3, 4]], 2, 4);
expected = [[1, 2], [3, 4]]
console.assert(expected.every(
    (row, r) =>
        row.every((item, c) =>
            item === result[r][c])
));
